perf/memtx.cc

```diff
@@ -0,0 +1,333 @@
+
+	struct space_def *space_def{};
+	struct key_def *key_def{};
+	struct index_def *tree_idx_def{};
+};
+
+/**
+ * The memtx fixture encapsulates the generation of the main key set used for
+ * benchmarking `select` and the key subset used for benchmarking `get`.
+ */
+class MemtxFixture : public benchmark::Fixture {
+public:
+	void
+	SetUp(const benchmark::State &) final
+	{
+		generate_key_set();
+		generate_key_subset();
+		empty_key.second = ::mp_encode_array(empty_key.first, 0);
+	}
+
+	void
+	TearDown(const benchmark::State &) final
+	{
+		remove_key_set();
+	}
+
+protected:
+	/**
+	 * Generate a key set of {uint} tuples and insert them into the index.
+	 */
+	void
+	generate_key_set()
+	{
+		for (std::size_t i = 0; i < key_set.size(); ++i) {
+			key_set[i].second =
+				::mp_encode_array(key_set[i].first, 1);
+			key_set[i].second =
+				::mp_encode_uint(key_set[i].second, i);
+			if (::box_insert(Memtx::sid, key_set[i].first,
+					 key_set[i].second, nullptr) != 0)
+				panic("failed to insert key");
+		}
+	}
+
+	/** Remove the key set by deleting all keys from the primary index. */
+	void
+	remove_key_set()
+	{
+		for (const auto &key : key_set) {
+			if (::box_delete(Memtx::sid, Memtx::primary_iid,
+					 key.first, key.second,
+					 nullptr) != 0)
+				panic("failed to delete key");
+		}
+	}
+
+	/**
+	 * Generate a key subset by choosing a random subset from the main key
+	 * set.
+	 */
+	void
+	generate_key_subset()
+	{
+		for(auto &k: key_subset) {
+			auto &key = key_set[distr(rng) % key_set.size()];
+			k = {key.first, key.second};
+		}
+	}
+
+	std::mt19937 rng{std::random_device()()};
+	std::uniform_int_distribution<std::mt19937::result_type> distr;
+
+	/** The key set used for benchmarking `select`. */
+	static constexpr std::size_t key_size_max = 5 + 9;
+	static constexpr std::size_t key_set_size = 1 << 17;
+	std::array<std::pair<char[key_size_max], char *>, key_set_size> key_set;
+
+	/** The key subset used for benchmarking `select`. */
+	static constexpr std::size_t key_subset_size = key_set_size / 2;
+	std::array<std::pair<char *, char *>, key_subset_size> key_subset;
+
+	/** The auxiliary empty key facilitates calling `select(nil)`. */
+	static constexpr std::uint32_t select_lim = 4294967295;
+	static constexpr std::size_t empty_key_size_max = 5;
+	std::pair<char[empty_key_size_max], char *> empty_key;
+};
+
+/**
+ * Benchmark random `get`s of existing keys from the tree index. The key subset
+ * is regenerated through every iteration. This benchmark is supposed to
+ * exercise the CPU and the cache.
+ */
+BENCHMARK_F(MemtxFixture, TreeGetRandomExistingKeys)
+(benchmark::State &state)
+{
+	struct tuple *t;
+	for (MAYBE_UNUSED auto _ : state) {
+		for (const auto &k: key_subset) {
+			benchmark::DoNotOptimize(::box_index_get(
+				Memtx::sid, Memtx::tree_iid, k.first, k.second,
+				&t));
+		}
+		state.PauseTiming();
+		generate_key_subset();
+		state.ResumeTiming();
+	}
+	state.SetItemsProcessed(key_subset_size);
```

Comment:
IIANM, `state.SetItemsProcessed()` value should be calculated this way:
```diff
--- a/perf/memtx.cc
+++ b/perf/memtx.cc
@@ -247,17 +247,19 @@ BENCHMARK_F(MemtxFixture, TreeGetRandomExistingKeys)
 (benchmark::State &state)
 {
        struct tuple *t;
+       size_t total_count = 0;
        for (MAYBE_UNUSED auto _ : state) {
                for (const auto &k: key_subset) {
                        benchmark::DoNotOptimize(::box_index_get(
                                Memtx::sid, Memtx::tree_iid, k.first, k.second,
                                &t));
+                       total_count++;
                }
                state.PauseTiming();
                generate_key_subset();
                state.ResumeTiming();
        }
-       state.SetItemsProcessed(key_subset_size);
+       state.SetItemsProcessed(total_count);
 }
```
Before this patch: `items_per_second=91.8609k/s`
After this patch: `items_per_second=1.53542M/s`

Reply:
Yeah, I made a typo in all `state.SetItemsProcessed` calls, fixed.